File: backend/agent/runtime/compression/estimation.py
from __future__ import annotations

from backend.agent.model.messages import Message, ToolResultBlock, ToolUseBlock


# Conservative chars-per-token estimate for English text
CHARS_PER_TOKEN = 4

# Overhead per message (role, separators, etc.)
MESSAGE_OVERHEAD_TOKENS = 4

# Overhead per tool block
TOOL_BLOCK_OVERHEAD_TOKENS = 8


def estimate_tokens(text: str) -> int:
    """Estimate token count for a text string.

    Uses a conservative estimate of ~4 chars per token.
    """
    if not text:
        return 0
    return max(1, len(text) // CHARS_PER_TOKEN)


def estimate_message_tokens(message: Message) -> int:
    """Estimate token count for a single Message.

    Includes message overhead and structured content.
    """
    total = MESSAGE_OVERHEAD_TOKENS

    content = message.content
    if isinstance(content, str):
        total += estimate_tokens(content)
    elif isinstance(content, list):
        for block in content:
            total += _estimate_block_tokens(block)

    return total


def _estimate_block_tokens(block: ToolUseBlock | ToolResultBlock) -> int:
    """Estimate tokens for a tool block."""
    total = TOOL_BLOCK_OVERHEAD_TOKENS

    if isinstance(block, ToolUseBlock):
        # Tool name
        total += estimate_tokens(block.name)
        # Tool input (serialized JSON)
        import json
        input_str = json.dumps(block.input, ensure_ascii=False)
        total += estimate_tokens(input_str)
    elif isinstance(block, ToolResultBlock):
        # Tool result content
        total += estimate_tokens(block.content)

    return total


def estimate_messages_tokens(messages: list[Message]) -> int:
    """Estimate total token count for a list of messages."""
    return sum(estimate_message_tokens(msg) for msg in messages)

File: server/src/agent/compression/estimation.ts
/**
 * Token estimation: conservative estimates with CJK character support.
 */

import type { LLMMessage, ContentBlock, ToolUseBlock, ToolResultBlock } from '@open-multi-agent/core'

/** Conservative chars-per-token for English/code. */
const CHARS_PER_TOKEN = 4

/** CJK characters typically use about 1 token per character. */
const CJK_PATTERN = /[\u3400-\u4DBF\u4E00-\u9FFF\uF900-\uFAFF]/g

/** Overhead per message (role, separators). */
const MESSAGE_OVERHEAD_TOKENS = 4

/** Overhead per tool block. */
const TOOL_BLOCK_OVERHEAD_TOKENS = 8

function countCJK(text: string): number {
  const matches = text.match(CJK_PATTERN)
  return matches ? matches.length : 0
}

export function estimateTokens(text: string): number {
  if (!text) return 0

  const englishEstimate = Math.floor(text.length / CHARS_PER_TOKEN)
  const cjkCount = countCJK(text)
  const nonCjkLen = text.length - cjkCount
  const cjkEstimate = cjkCount + Math.floor(nonCjkLen / CHARS_PER_TOKEN)

  return Math.max(1, Math.max(englishEstimate, cjkEstimate))
}

function estimateBlockTokens(block: ContentBlock): number {
  let total = TOOL_BLOCK_OVERHEAD_TOKENS

  if (block.type === 'tool_use') {
    const tb = block as ToolUseBlock
    total += estimateTokens(tb.name)
    total += estimateTokens(JSON.stringify(tb.input, null, 0))
  } else if (block.type === 'tool_result') {
    const rb = block as ToolResultBlock
    total += estimateTokens(rb.content)
  } else if (block.type === 'text') {
    total += estimateTokens((block as any).text ?? '')
  } else if (block.type === 'reasoning') {
    total += estimateTokens((block as any).thinking ?? '')
  }

  return total
}

export function estimateMessageTokens(message: LLMMessage): number {
  let total = MESSAGE_OVERHEAD_TOKENS

  if (Array.isArray(message.content)) {
    for (const block of message.content) {
      total += estimateBlockTokens(block)
    }
  }

  return total
}

export function estimateMessagesTokens(messages: LLMMessage[]): number {
  let sum = 0
  for (const msg of messages) {
    sum += estimateMessageTokens(msg)
  }
  return sum
}
